internal/servers/worker/internal/metric/proxy_ws.go

```diff
@@ -0,0 +1,155 @@
+package metric
+
+import (
+	"bufio"
+	"net"
+	"net/http"
+	"sync"
+
+	"github.com/hashicorp/boundary/globals"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	proxyWebsocketSubsystem = "worker_proxy_websocket"
+)
+
+var (
+	activeConnCount = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: globals.MetricNamespace,
+			Subsystem: proxyWebsocketSubsystem,
+			Name:      "active_connections",
+			Help:      "Count of open websocket proxy connections (to Boundary workers).",
+		},
+	)
+
+	bytesReceived = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: globals.MetricNamespace,
+			Subsystem: proxyWebsocketSubsystem,
+			Name:      "received_bytes_total",
+			Help:      "Count of received bytes for Worker proxy websocket connections.",
+		},
+	)
+
+	bytesSent = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: globals.MetricNamespace,
+			Subsystem: proxyWebsocketSubsystem,
+			Name:      "sent_bytes_total",
+			Help:      "Count of sent bytes for Worker proxy websocket connections.",
+		},
+	)
+)
+
+// InitializeWebsocketCollectors registers the websocket collectors onto `r`.
+// It panics upon the first registration that causes an error.
+func InitializeWebsocketCollectors(r prometheus.Registerer) {
+	if r == nil {
+		return
+	}
+	r.MustRegister(activeConnCount, bytesReceived, bytesSent)
+}
+
+// InstrumentWebsocketWrapper expects an http.Handler where websockets are used
+// and wraps it to instrument that websocket with metrics (bytes received, sent, etc).
+func InstrumentWebsocketWrapper(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		nextRw := rw
+		if rwh, ok := rw.(http.Hijacker); ok {
+			nextRw = wrapProxyHandler(&hijackerWrapper{ResponseWriter: rw, Hijacker: rwh}, rw)
+		}
+		next.ServeHTTP(nextRw, req)
+	})
+}
+
+func wrapProxyHandler(with *hijackerWrapper, wrap http.ResponseWriter) http.ResponseWriter {
```

Comment:
Since it wasn't immediately intuitive why using this function is helpful, consider adding an explanation for why this function exist (so downstream code can cast to these interfaces).

Reply:
[5954753](https://github.com/hashicorp/boundary/pull/1963/commits/59547537e4a5be499d98f5d521b102ab2e057702)